### PR TITLE
Fix argc bug #127 in parser and adddocumentation

### PR DESCRIPTION
**Pull request by Superlokkus**
Hi Vladimir,

as discussed in #127 . Also added documentation, with the intention to avoid regression or stupid questions like mine in the future :-) .

### CODE REVIEW

File: include/boost/program_options/parsers.hpp [hunk 1]
```diff
@@ -120,7 +120,10 @@ namespace boost { namespace program_options {
         basic_command_line_parser(const std::vector<
                                   std::basic_string<charT> >& args);
         /** Creates a command line parser for the specified arguments
-            list. The parameters should be the same as passed to 'main'.
+            list. The parameters should be the same as passed to 'main', meaning:
+            @param argc Must be non-zero i.e. >= 0
```
Comment (sehe) [inline include/boost/program_options/parsers.hpp:124]: Two things. Non-zero is >0. Secondly, I think the standard mandates a argc>=1 and argv[0] is the program name on all known platforms

File: include/boost/program_options/parsers.hpp [hunk 2]
```diff
@@ -120,7 +120,10 @@ namespace boost { namespace program_options {
...
+            @param argv argv[0] up to argv[argc-1] must point to null terminated strings and
+            argv[argc] must be 0 e.g. nullptr
```
Comment (sehe) [inline include/boost/program_options/parsers.hpp:126]: why was this added here? It appears as though this restriction is never relied on here (in fact neither `argc` or `argv` are in sight). Perhaps this part of POSIX semantics should be documented elsewhere?

File: include/boost/program_options/parsers.hpp (hunk 1 above)
Comment (Superlokkus) [reply include/boost/program_options/parsers.hpp:124]: I just had a discussion about that prior with a theoretical physicist: according to him, everyone in the math world defines non-zero with >=0. Also got references: https://mathworld.wolfram.com/Nonnegative.html 

Math world besides, as mentioned in bug report, cpp standard seem to share this definition too. As it defines the behavior for argc == 0 cases.

File: include/boost/program_options/parsers.hpp (hunk 2 above)
Comment (Superlokkus) [reply include/boost/program_options/parsers.hpp:126]: I did not refer to posix but the [C++20 draft N 4860](https://isocpp.org/files/papers/N4860.pdf) and https://eel.is/c++draft/basic.start.main#2.2 in the bug report: 
To quote 6.9.3.1 main function [basic.start.main] (2.2) from the draft

> If argc is nonzero these arguments shall be supplied in argv[0] through argv[argc-1] as pointers to the initial characters of null-terminated multibyte strings (ntmbss) (16.4.2.2.5.2) and argv[0] shall be the pointer to the initial character of a ntmbs that represents the name used to invoke the program or "". The value of argc shall be non-negative. The value of argv[argc] shall be 0.

or in [the newer one](https://eel.is/c++draft/basic.start.main#2.2)

So this comment shall make a regression of that bug less likely and state an API contract for function users and boost code contributors, what to expect.

File: include/boost/program_options/parsers.hpp (hunk 1 above)
Comment (Superlokkus) [reply include/boost/program_options/parsers.hpp:124]: To the second point, see comment below

Comment (sehe) [reply include/boost/program_options/parsers.hpp:124]: > Also got references: https://mathworld.wolfram.com/Nonnegative.html

That's Nonnegative, though

File: include/boost/program_options/parsers.hpp (hunk 2 above)
Comment (sehe) [reply include/boost/program_options/parsers.hpp:126]: Ah. My confusion was due to github UI putting the comment "together" with the preceding constructor overload. I was struggling to see how it related there :) Disregard

File: include/boost/program_options/parsers.hpp (hunk 1 above)
Comment (Superlokkus) [reply include/boost/program_options/parsers.hpp]: Resolved?

Comment (sehe) [reply include/boost/program_options/parsers.hpp]: Sure (I cannot find a button to mark as resolved)

Comment (Superlokkus) [reply include/boost/program_options/parsers.hpp]: Well for some reason I can resolve it. But this of course only makes sense, if you would have :-)

**Comment by CrazyPandas** [comment]
I found that this issue was mentioned as early as 2018.

https://github.com/boostorg/program_options/pull/59

**Review by vprus — APPROVED** [review]

**Comment by vprus** [comment]
Thanks, this PR was merged.
